edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnection.cs

```diff
@@ -258,38 +259,64 @@ void InternalConnectionStatusChangesHandler(ConnectionStatus status, ConnectionS
         /// If the existing identity has a usable token, then use it.
         /// Else, generate a notification of token being near expiry and return a task that
         /// can be completed later.
+        /// Keep retrying till we get a usable token.
         /// Note - Don't use this.Identity in this method, as it may not have been set yet!
         /// </summary>
         async Task<string> GetNewToken(string iotHub, string id, string currentToken, IIdentity currentIdentity)
         {
             Events.GetNewToken(id);
-            // We have to catch UnauthorizedAccessException, because on IsTokenUsable, we call parse from
-            // Device Client and it throws if the token is expired.
-            if (IsTokenUsable(iotHub, currentToken))
-            {
-                Events.UsingExistingToken(id);
-                return currentToken;
-            }
-            else
+            bool retrying = false;
+            string token = currentToken;
+            while (true)
             {
-                Events.TokenExpired(id, currentToken);
-            }
+                // We have to catch UnauthorizedAccessException, because on IsTokenUsable, we call parse from
+                // Device Client and it throws if the token is expired.
+                if (IsTokenUsable(iotHub, token))
+                {
+                    if (retrying)
+                    {
+                        Events.NewTokenObtained(iotHub, id, token);
+                    }
+                    else
+                    {
+                        Events.UsingExistingToken(id);
+                    }
+                    return token;
+                }
+                else
+                {
+                    Events.TokenNotUsable(iotHub, id, token);                    
+                }
 
-            // No need to lock here as the lock is being held by the refresher.
-            TaskCompletionSource<string> tcs = this.tokenGetter
-                .GetOrElse(
-                    () =>
+                bool newTokenGetterCreated = false;
+                // No need to lock here as the lock is being held by the refresher.
+                TaskCompletionSource<string> tcs = this.tokenGetter
+                    .GetOrElse(
+                        () =>
+                        {
+                            Events.SafeCreateNewToken(id);
+                            var taskCompletionSource = new TaskCompletionSource<string>();
+                            this.tokenGetter = Option.Some(taskCompletionSource);
+                            newTokenGetterCreated = true;                            
+                            return taskCompletionSource;
+                        });
+
+                // If a new tokenGetter was created, then invoke the connection status changed handler
```

Comment:
It is still not clear to me from the comments when a new tokenGetter is created. That is key to understanding this loop.

Reply:
Given the lack of general documentation in our code base, having comments that explain not just things local to a piece of code, but spanning components of code, would be very helpful.

Reply:
Yes, I agree with you. I am just a bit weary of comments getting stale in that case...